src/operator/setup.rs

```diff
@@ -0,0 +1,185 @@
+use crate::test::prelude::{Node, Pod, TestKubeClient};
+
+use anyhow::{anyhow, Result};
+use kube::api::ObjectList;
+use kube::{Resource, ResourceExt};
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use stackable_operator::status::Conditions;
+use std::fmt::Debug;
+use std::thread;
+use std::time::{Duration, Instant};
+
+/// A Wrapper to avoid passing in client or cluster everywhere.
+pub struct TestCluster<T> {
+    client: TestKubeClient,
+    cluster: Option<T>,
+    options: TestClusterOptions,
+    timeouts: TestClusterTimeouts,
+}
+
+/// Some reoccurring common test cluster options.
+pub struct TestClusterOptions {
+    pub cluster_ready_condition_type: String,
+    pub pod_name_label: String,
+}
+
+/// Some reoccurring common test cluster timeouts.
+pub struct TestClusterTimeouts {
+    pub cluster_ready: Duration,
+    pub pods_terminated: Duration,
+}
+
+impl<T> TestCluster<T>
+where
+    T: Clone + Debug + DeserializeOwned + Resource<DynamicType = ()> + Serialize,
+{
+    /// This creates a kube client and should be executed at the start of every test.
+    pub fn new(options: TestClusterOptions, timeouts: TestClusterTimeouts) -> Self {
+        TestCluster {
+            client: TestKubeClient::new(),
+            cluster: None,
+            options,
+            timeouts,
+        }
+    }
+
+    /// Applies a custom resource, stores the returned cluster object and sleeps for
+    /// to seconds to give the operator time to react on the custom resource.
+    /// Without the sleep it can happen that tests run without any pods being created.
+    fn apply(&mut self, cluster: &T) -> Result<()> {
+        self.cluster = Some(self.client.apply(&serde_yaml::to_string(cluster)?));
+
+        // we wait here to give the operator time to react to the custom resource
+        thread::sleep(Duration::from_secs(2));
+        Ok(())
+    }
+
+    /// Applies a command and waits 2 seconds to let the operator react on in.
+    pub fn apply_command<C>(&self, command: &C) -> Result<C>
+    where
+        C: Clone + Debug + DeserializeOwned + Resource<DynamicType = ()> + Serialize,
+    {
+        let cmd: C = self.client.apply(&serde_yaml::to_string(command)?);
+
+        // we wait here to give the operator time to react to the command
+        thread::sleep(Duration::from_secs(2));
+        Ok(cmd)
+    }
+
+    /// Deletes the custom resource, and waits for pods to be terminated
+    /// This should be executed after every single test to provide a "clean"
+    /// environment for the following tests.
```

Comment:
If an assertion fails then the thread panics and no cleanup code is run. Therefore it could be better to provide a `Drop` implementation or wrap this structure in a `TemporaryResource`.

Reply:
I added the drop trait, so i will delete the "manual" cleanups delete and delete_commands.